### PR TITLE
Marshal.load: do not call the proc until strings have their encoding

### PR DESCRIPTION
**Pull request by byroot**
Ref: https://bugs.ruby-lang.org/issues/18141

### CODE REVIEW

File: spec/ruby/core/marshal/shared/load.rb [hunk 1]
```diff
@@ -20,6 +20,24 @@
   end
 
   describe "when called with a proc" do
+    ruby_version_is '3.1' do
```
Comment (nobu) [inline spec/ruby/core/marshal/shared/load.rb:23]:
```suggestion
    ruby_bug "#18141", ""..."3.1" do
```

File: spec/ruby/core/marshal/shared/load.rb [hunk 2]
```diff
@@ -35,27 +53,54 @@
...
+        Marshal.send(@method, "\x04\bI[\vI\"\ahi\a:\x06EF:\t@fooi\ni\x0F@\x06@\x06IS:\x14Struct::Brittle\x06:\x06af\x060\x06:\n@clueI\"\tnone\x06;\x00FI[\b;\b:\x06b:\x06c\x06:\t@twoi\a\x06:\t@zooI\"\bant\x06;\x00F", proc)
+
+        arr.should == [false, 5, "hi", 10, "hi", "hi", 0.0, st, false, "none",
+                       :b, :c, a, 2, ["hi", 10, "hi", "hi", st, [:a, :b, :c]], false, "ant"]
+
+        Struct.send(:remove_const, :Brittle)
+      end
+    end
+
+    ruby_bug "#18141", ""..."3.1"
```
Comment (byroot) [inline spec/ruby/core/marshal/shared/load.rb:81]: @nobu I edited like you suggested.

File: spec/ruby/core/marshal/shared/load.rb [hunk 3]
```diff
@@ -35,27 +53,54 @@
...
-      obj.instance_variable_set(:@zoo, 'ant')
-      proc = Proc.new { |o| arr << o; o}
-
-      Marshal.send(@method, "\x04\bI[\vI\"\ahi\a:\x06EF:\t@fooi\ni\x0F@\x06@\x06IS:\x14Struct::Brittle\x06:\x06af\x060\x06:\n@clueI\"\tnone\x06;\x00FI[\b;\b:\x06b:\x06c\x06:\t@twoi\a\x06:\t@zooI\"\bant\x06;\x00F", proc)
-
-      arr.should == ["hi", false, 5, 10, "hi", "hi", 0.0, st, "none", false,
-                     :b, :c, a, 2, ["hi", 10, "hi", "hi", st, [:a, :b, :c]], "ant", false]
-
-      Struct.send(:remove_const, :Brittle)
+    ruby_version_is '3.1' do
```
Comment (nobu) [inline spec/ruby/core/marshal/shared/load.rb:56]:
```suggestion
    ruby_bug "#18141", ""..."3.1" do
```

File: spec/ruby/core/marshal/shared/load.rb [hunk 4]
```diff
@@ -35,27 +53,54 @@
...
+        obj.instance_variable_set(:@zoo, 'ant')
+        proc = Proc.new { |o| arr << o; o}
+
+        Marshal.send(@method, "\x04\bI[\vI\"\ahi\a:\x06EF:\t@fooi\ni\x0F@\x06@\x06IS:\x14Struct::Brittle\x06:\x06af\x060\x06:\n@clueI\"\tnone\x06;\x00FI[\b;\b:\x06b:\x06c\x06:\t@twoi\a\x06:\t@zooI\"\bant\x06;\x00F", proc)
+
+        arr.should == ["hi", false, 5, 10, "hi", "hi", 0.0, st, "none", false,
+                       :b, :c, a, 2, ["hi", 10, "hi", "hi", st, [:a, :b, :c]], "ant", false]
+
+        Struct.send(:remove_const, :Brittle)
+      end
```
Comment (nobu) [inline spec/ruby/core/marshal/shared/load.rb:103]:
```suggestion
```

**Review by nobu — CHANGES_REQUESTED** [review]
This is a bug fix, not a feature change.

**Comment by byroot** [comment]
The Windows build failed, not sure why. All other builds are green.

File: spec/ruby/core/marshal/shared/load.rb [hunk 5]
```diff
@@ -20,6 +20,24 @@
...
+            arg.should == utf8_string
+            arg.encoding.should == Encoding::UTF_8
+          end
+          arg
+        })
+      end
+
+      it "no longer mutate the object after it was passed to the proc" do
+        string = Marshal.load(Marshal.dump("foo"), :freeze.to_proc)
+        string.frozen?.should == true
```
Comment (nobu) [inline spec/ruby/core/marshal/shared/load.rb:37]:
```suggestion
        string.should.frozen?
```
This is a newer style, and will show a better (or more descriptive) message at failure.

**Review by nobu — APPROVED** [review]
Why are two `ruby_bug` blocks separated?

**Comment by byroot** [comment]
@nobu thanks for the review, I fixed both your comments.

**Comment by nobu** [comment]
Replacing "loads an Array with proc"  block makes the change clear by `diff -w`, I guess.

**Comment by byroot** [comment]
> Replacing "loads an Array with proc" block makes the change clear by `diff -w`, I guess.

Not sure I understand what you mean. Are you suggesting further changes to the PR?

**Comment by nobu** [comment]
> Are you suggesting further changes to the PR?

Sorry, yes.

**Comment by byroot** [comment]
What change are you suggesting?

**Comment by nobu** [comment]
I mean, moving the added code to the deleted place.

**Comment by byroot** [comment]
@nobu done.

**Pull request merged**